pandas/_libs/tslib.pyx

```diff
@@ -595,6 +633,30 @@ cpdef array_to_datetime(ndarray[object] values, errors='raise',
                             continue
                         raise TypeError("invalid string coercion to datetime")
 
+                    # If the dateutil parser returned tzinfo, capture it
+                    # to check if all arguments have the same tzinfo
+                    tz = py_dt.tzinfo
+                    if tz is not None:
+                        seen_datetime_offset = 1
+                        if tz == dateutil_utc():
+                            # dateutil.tz.tzutc has no offset-like attribute
+                            # Just add the 0 offset explicitly
+                            out_tzoffset_vals.add(0)
+                        elif tz == tzlocal():
+                            # is comparison fails unlike other dateutil.tz
+                            # objects. Also, dateutil.tz.tzlocal has no
+                            # _offset attribute like tzoffset
+                            offset_seconds = tz._dst_offset.total_seconds()
```

Comment:
Note that this is using an internal method and is fragile.

Reply:
Good point.  Suggestion for a better way to do this?

Reply:
I'm not sure I understand what this function is supposed to do, but from what I *do* understand it looks like a mistake.

Reply:
My original intent with this section was to collect the dateutil timezone objects in a `set` and determine if more than 1 distinct timezone object was collected.

However since dateutil timezones cannot be hashed (thanks for starting progress on the associated issue btw @pganssle), I instead store the timezone's UTC offsets in seconds instead (which is what I essentially care about).

It wasn't too apparent to me if there was a public way to access a dateutil timezone's UTC offset, hence why I am use a private method here. Definitely open to a better method.

Reply:
I think if you can't find a public method you probably shouldn't dive into the private API. There are a few assumptions in here that won't survive future versions of `dateutil` for sure. The `tzlocal` thing also actually gives you the *wrong answer*, because `tzlocal` is not a fixed offset and you're taking the DST offset.

If you want the offset, you should do `dt.utcoffset()`, regardless of whether it's `dateutil` or not.

Reply:
If you are significantly worried about speed, you can probably use the fact that in `dateutil >= 2.7.0`, `dateutil.tz.tzutc()` returns a singleton, and `dateutil.tz.tzoffset(*args1)` is the same object as any other `dateutil.tz.tzoffset(*args2)` where `args1 == args2`. So you can probably store a mapping between `id(obj)` and the result of `obj.tzoffset(datetime(1970, 1, 1))` for that specific subclass (since it is also guaranteed to have no DST).